pkg/kubevirt/core/core.go

```diff
@@ -0,0 +1,320 @@
+		if err != nil {
+			return "", fmt.Errorf("invalid dns policy: %v", err)
+		}
+	}
+
+	if providerSpec.DNSConfig != "" {
+		if err := yaml.Unmarshal([]byte(providerSpec.DNSConfig), dnsConfig); err != nil {
+			return "", fmt.Errorf(`failed to unmarshal "dnsConfig" field: %v`, err)
+		}
+	}
+
+	virtualMachine := &kubevirtv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      machineName,
+			Namespace: providerSpec.Namespace,
+			Labels: map[string]string{
+				"kubevirt.io/vm": machineName,
+			},
+		},
+		Spec: kubevirtv1.VirtualMachineSpec{
+			Running: utilpointer.BoolPtr(true),
+			Template: &kubevirtv1.VirtualMachineInstanceTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"kubevirt.io/vm": machineName,
+					},
+				},
+				Spec: kubevirtv1.VirtualMachineInstanceSpec{
+					Domain: kubevirtv1.DomainSpec{
+						Devices: kubevirtv1.Devices{
+							Disks: []kubevirtv1.Disk{
+								{
+									Name:       "datavolumedisk",
+									DiskDevice: kubevirtv1.DiskDevice{Disk: &kubevirtv1.DiskTarget{Bus: "virtio"}},
+								},
+								{
+									Name:       "cloudinitdisk",
+									DiskDevice: kubevirtv1.DiskDevice{Disk: &kubevirtv1.DiskTarget{Bus: "virtio"}},
+								},
+							},
+						},
+						Resources: kubevirtv1.ResourceRequirements{
+							Requests: *requestsAndLimits,
+							Limits:   *requestsAndLimits,
+						},
+					},
+					TerminationGracePeriodSeconds: &terminationGracePeriodSeconds,
+					Volumes: []kubevirtv1.Volume{
+						{
+							Name: "datavolumedisk",
+							VolumeSource: kubevirtv1.VolumeSource{
+								DataVolume: &kubevirtv1.DataVolumeSource{
+									Name: machineName,
+								},
+							},
+						},
+						{
+							Name: "cloudinitdisk",
+							VolumeSource: kubevirtv1.VolumeSource{
+								CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+									UserDataSecretRef: &corev1.LocalObjectReference{
+										Name: userdataSecretName,
+									},
+								},
+							},
+						},
+					},
+					DNSPolicy: dnsPolicy,
+					DNSConfig: dnsConfig,
+				},
+			},
+			DataVolumeTemplates: []cdi.DataVolume{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: machineName,
+					},
+					Spec: cdi.DataVolumeSpec{
+						PVC: &corev1.PersistentVolumeClaimSpec{
+							StorageClassName: utilpointer.StringPtr(providerSpec.StorageClassName),
+							AccessModes: []corev1.PersistentVolumeAccessMode{
+								"ReadWriteOnce",
+							},
+							Resources: corev1.ResourceRequirements{
+								Requests: pvcRequest,
+							},
+						},
+						Source: cdi.DataVolumeSource{
+							HTTP: &cdi.DataVolumeSourceHTTP{
+								URL: providerSpec.SourceURL,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if p.client == nil {
+		kubevirtClient, err := kubevirtClient(secrets)
+		if err != nil {
+			return "", fmt.Errorf("failed to get kubevirt client: %v", err)
+		}
+		p.client = kubevirtClient
+	}
```

Comment:
This means you are actually relying on the constructor completely and any change in the secret wouldn't matter at all(only on the provider initialising). If this ok for your architecture then sure we can do that. But then passing those secrets, which to the methods won't make any sense(which is following the driver interface).

Reply:
In your existing code, if the secrets change, this will also not have any effect since you are saving the created client in the `client` field, so any subsequent calls with a different secret would use the already created client.
Regarding whether the secret can change: as far as I understand, when creating the client, you are only using the `kubeconfig` for your "provider" cluster from the secrets, correct? Can this "provider" cluster change during the lifetime of a shoot? I don't think so, but to be on the safe side, we should probably discuss it.
To be on the safe side, and having in mind that the secrets are passed in, my initial approach of injecting the client should be changed (and it's actually not feasible since without the secrets you can't create it, and at the point in time of calling the constructor you don't have the secrets, only during an actual method call). So therefore, you could inject a `ClientFactory` - an interface that has a single `NewClient(kubeconfig []byte)` method. The actual implementation of this factory would create the client. Note that you should not in this case save the result in the field, just use the created client only within the method it was created.